### PR TITLE
Add Ballot Leader Election module and a Memory Storage

### PR DESCRIPTION
**Pull request by Nightcro**
Please make sure these boxes are checked, before submitting a new PR.

- [x] You ran `rustfmt` on the code base before submitting (on latest nightly with rustfmt support)
- [x] You reference which issue is being closed in the PR text (if applicable)

## Breaking Changes

Fixed linter warning: Decide.with() - order of parameters changed
In Paxos module get_decided_entries was renamed to get_last_decided_entries.

## Other Changes

Ported Ballot Leader Election using tick() as a timer.
Ported MemoryStorage

Signed-off-by: Nightcro <alexandru20059@gmail.com>

### CODE REVIEW

File: src/ble/ballot_leader_election.rs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+/// incoming messages and produces outgoing messages that the user has to fetch periodically and send using a network implementation.
+/// User also has to periodically fetch the decided entries that are guaranteed to be strongly consistent and linearizable, and therefore also safe to be used in the higher level application.
+pub struct BallotLeaderComp {
+    pid: u64,
+    pub(crate) peers: Vec<u64>,
+    hb_round: u32,
+    ballots: Vec<(Ballot, bool)>,
+    current_ballot: Ballot, // (round, pid)
+    candidate: bool,
+    leader: Option<Ballot>,
```
Comment (haraldng) [inline src/ble/ballot_leader_election.rs:32]: I think the user should be able to read who is the current leader at any time. So let's create a public function `get_leader()` that returns this.

**Review by haraldng — CHANGES_REQUESTED** [review]
Most of it is about docs and naming. However, please take a closer look at `tick()` again (or let me know if I have misunderstood the intended usage for it).

**Review by haraldng — CHANGES_REQUESTED** [review]
Some more comments on docs and some small issues in BLE.

File: src/leader_election.rs [hunk 2]
```diff
@@ -25,3 +25,342 @@ where
...
+            self.ticks_elapsed += 1;
+
+            if self.ticks_elapsed >= self.hb_current_delay {
+                self.ticks_elapsed = 0;
+                self.hb_timeout()
+            } else {
+                None
+            }
+        }
+
```
Comment (haraldng) [inline src/leader_election.rs:167]: We must have a public function that returns the outgoing messages. We implemented one in the integration tests. The documentation should be similar to tick() that asks the user to call it repeatedly.

**Comment by Nightcro** [comment]
Fixes #11

**Review by haraldng — CHANGES_REQUESTED** [review]

File: src/leader_election.rs [hunk 3]
```diff
@@ -70,20 +70,18 @@ pub mod ballot_leader_election {
         majority_connected: bool,
         /// Current elected leader.
         leader: Option<Ballot>,
-        /// Delay until timeout.
-        hb_current_delay: u64,
+        /// Internal delay used for timeout.
+        _hb_current_delay: u64,
```
[the remark this answers is not in the document]
Comment (Nightcro) [reply src/leader_election.rs:74]: I wasn't sure, how to state that this was an internal variable, not to be seen by the user
I can remove the underscore

[the remark this answers is not in the document]
Comment (haraldng) [reply src/leader_election.rs:74]: Yes, remove the underscore. I think it should be private by default if you don't have `pub` before the variable

**Pull request merged**